lib/route/cls/flower.c

```diff
@@ -0,0 +1,779 @@
+	if (err < 0)
+		return err;
+
+	if (tb[TCA_FLOWER_FLAGS]) {
+		f->cf_flags = nla_get_u32(tb[TCA_FLOWER_FLAGS]);
+		f->cf_mask |= FLOWER_ATTR_FLAGS;
+	}
+
+	if (tb[TCA_FLOWER_ACT]) {
+		err = rtnl_act_parse(&f->cf_act, tb[TCA_FLOWER_ACT]);
+		if (err)
+			return err;
+
+		f->cf_mask |= FLOWER_ATTR_ACTION;
+	}
+
+	if (tb[TCA_FLOWER_POLICE]) {
+		f->cf_police = nl_data_alloc_attr(tb[TCA_FLOWER_POLICE]);
+		if (!f->cf_police)
+			return -NLE_NOMEM;
+
+		f->cf_mask |= FLOWER_ATTR_POLICE;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_TYPE]) {
+	        f->cf_proto = nla_get_u16(tb[TCA_FLOWER_KEY_ETH_TYPE]);
+		f->cf_mask |= FLOWER_ATTR_PROTO;
+	}
+
+	if (tb[TCA_FLOWER_KEY_VLAN_ID]) {
+	        f->cf_vlan_id = nla_get_u16(tb[TCA_FLOWER_KEY_VLAN_ID]);
+		f->cf_mask |= FLOWER_ATTR_VLAN_ID;
+	}
+
+	if (tb[TCA_FLOWER_KEY_VLAN_PRIO]) {
+	        f->cf_vlan_prio = nla_get_u16(tb[TCA_FLOWER_KEY_VLAN_PRIO]);
+		f->cf_mask |= FLOWER_ATTR_VLAN_PRIO;
+	}
+
+	if (tb[TCA_FLOWER_KEY_VLAN_ETH_TYPE]) {
+	        f->cf_vlan_ethtype = nla_get_u16(tb[TCA_FLOWER_KEY_VLAN_ETH_TYPE]);
+		f->cf_mask |= FLOWER_ATTR_VLAN_ETH_TYPE;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_DST]) {
+	        f->cf_dst_mac = nl_data_alloc_attr(tb[TCA_FLOWER_KEY_ETH_DST]);
+		if (!f->cf_dst_mac)
+		        return -NLE_NOMEM;
+
+		f->cf_mask |= FLOWER_ATTR_DST_MAC;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_DST_MASK]) {
+	        f->cf_dst_mac_mask = nl_data_alloc_attr(tb[TCA_FLOWER_KEY_ETH_DST_MASK]);
+		if (!f->cf_dst_mac_mask)
+		        return -NLE_NOMEM;
+
+		f->cf_mask |= FLOWER_ATTR_DST_MAC_MASK;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_SRC]) {
+	        f->cf_src_mac = nl_data_alloc_attr(tb[TCA_FLOWER_KEY_ETH_SRC]);
+		if (!f->cf_src_mac)
+		        return -NLE_NOMEM;
+
+		f->cf_mask |= FLOWER_ATTR_SRC_MAC;
+	}
+
+	if (tb[TCA_FLOWER_KEY_ETH_SRC_MASK]) {
+	        f->cf_src_mac_mask = nl_data_alloc_attr(tb[TCA_FLOWER_KEY_ETH_SRC_MASK]);
+		if (!f->cf_src_mac_mask)
+		        return -NLE_NOMEM;
+
+		f->cf_mask |= FLOWER_ATTR_SRC_MAC_MASK;
+	}
+
+	if (tb[TCA_FLOWER_KEY_IP_TOS]) {
+	        f->cf_ip_dscp = nla_get_u8(tb[TCA_FLOWER_KEY_IP_TOS]);
+		f->cf_mask |= FLOWER_ATTR_IP_DSCP;
+	}
+
+	if (tb[TCA_FLOWER_KEY_IP_TOS_MASK]) {
+	        f->cf_ip_dscp_mask = nla_get_u8(tb[TCA_FLOWER_KEY_IP_TOS_MASK]);
+		f->cf_mask |= FLOWER_ATTR_IP_DSCP_MASK;
+	}
+
+        return 0;
+}
+
+static int flower_msg_fill(struct rtnl_tc *tc, void *data, struct nl_msg *msg)
+{
+        struct rtnl_flower *f = data;
+	int err;
+
+	if (!f)
+	        return 0;
+
+	if (f->cf_mask & FLOWER_ATTR_FLAGS)
+	        NLA_PUT_U32(msg, TCA_FLOWER_FLAGS, f->cf_mask);
+
+	if (f->cf_mask & FLOWER_ATTR_ACTION) {
+	        err = rtnl_act_fill(msg, TCA_FLOWER_ACT, f->cf_act);
+		if (err)
+		        return err;
+	}
+
+	if (f->cf_mask & FLOWER_ATTR_POLICE)
+	        NLA_PUT_DATA(msg, TCA_FLOWER_POLICE, f->cf_police);
+
+	if (f->cf_mask & FLOWER_ATTR_PROTO)
+	        NLA_PUT_U16(msg, TCA_FLOWER_KEY_ETH_TYPE, f->cf_proto);
```

Comment:
this and most (all?) other U16 values need to be in big endian byte order on netlink API. As I suggested above to correct for endianness during parsing, the process must be reverted during `NLA_PUT_U16()`.

Reply:
cf_proto and cf_vlan_ethtype are always in big-endian. No conversion is needed here.